### PR TITLE
Implementation of a gulag style mini-game

### PR DESCRIPTION
**Pull request by derDeidra**
This pull request adds a gulag style mini-game to the d.gg chat.  The idea behind the game is that users can be sent to the gulag and forced to battle it out against other prisoners for a chance at freedom.  Winners will be celebrated and losers will be muted.  Matches occur at a configurable interval while there are prisoners in the gulag.  The prisoners in the match are asked a question and the first prisoner to respond with the correct answer will win the match.

The first pass just includes randomized math questions, other random trivia stuff can easily be added.

This was inspired by a shitpost on the subreddit.  I thought it would be a fun meme.  Let me know if this looks like something that would be fun to have in chat.

It's also worth noting that this initial implementation was done based on a very rudimentary look over the bot codebase, its quite possible there are better ways to accomplish what I'm trying to do within the existing framework.

### CODE REVIEW

**Comment by coolcodehere** [comment]
Great commit!

**Comment by zoulos** [comment]
can u add rock throwing.........

**Comment by niceprogramming** [comment]
gulags are not funny 😠😠😠😠😠😠😠😠😠😠

**Comment by kevinchoinj** [comment]
rpi 🎊

**Comment by shadedred** [comment]
Let's discuss the merits of this

**Comment by ILiedAboutCake** [comment]
this should be its own command to send 2 users to the gulag, and not tied to global ban or nuke commands used for  ~~moderation~~ (raging in league games)

Also this should be able to be disabled in the config file

**Comment by derDeidra** [comment]
Hi folks, I cleaned up the implementation a little bit.  The commands and scheduler scripts are now totally self contained and do not modify any of the existing moderation tools.  The game supports a few different commands:

* gulag -> privileged command, consign user(s) to the gulag queue
* release -> privileged command, release a user from the gulag queue
* purge -> privileged command, functions similarly to nuke, adds all users who used a 'purged' phrase to the gulag (does not do any of the nuke caching stuff to catch subsequent usages for now)
* revolution -> privileged command, release all users from the gulag queue
* answer -> non-privileged command, used by gulag participants to answer the match question
* stone -> non-privileged command, used by anyone to knock out the answer of a gulag participant, forcing them re-answer the question

I also added a configuration option to disable the gulag scheduler.  If you want me to integrate the disable option to block the registration of the commands as well, just let me know.

**Comment by coolcodehere** [comment]
Thanks for all your hard work!

**Comment by shadedred** [comment]
im down to try it out @ILiedAboutCake , could be a good meme. I don't think we'd ever use the match scheduler itself, that might spam up chat, but using the manually started matches could be a good meme

**Comment by johnpyp** [comment]
Thanks for all your hard work!

**Comment by johnpyp** [comment]
These commits are truly more numerous than beverly's :)

**Comment by derDeidra** [comment]
@shadedred Ah let me clarify, when is say match scheduler I do not mean that there is an automated process that randomly gulags people.  The match scheduler is what pairs people off once they are placed into the gulag via some moderator action.  I.e. if you nuke a phrase and it places 10 people into the queue, the scheduler handles running 5 matches one by one and pairing the opponents up.

**Comment by shadedred** [comment]
That definitely might be too much spam, as in the chat would just be people gulaging for like a solid minute or two. I think a single 1v1 instance of a gulag event only, would be more manageable. waiting for Cake to weight in, he's got the final say

**Comment by derDeidra** [comment]
@shadedred Yeah for sure you probably wouldn't want to gulag that many people at one time since the matches would go on for a while.  I just wanted to clarify the behavior of the match scheduler and alley any concerns about automated gulaging.  All matches occur at the discretion of the moderation team.

**Comment by ILiedAboutCake** [comment]
Everything looks great so far just doing a code review, can you update the commands documentation in the readme? Also, are there any tests that can be written for this that we don't already have?

**Comment by derDeidra** [comment]
@ILiedAboutCake I have updated the README and added a suite of tests.  Let me know if everything else looks good.

File: tests/lib/services/gulag-runner.test.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+  });
+
+  it('can handle running a gulag match with no winner where both players answer incorrectly', (done) => {
+    this.runner.countdownHandle = 'dummyhandle';
+    this.runner.addPrisoner('prisoner1');
+    this.runner.addPrisoner('prisoner2');
+    this.runner.matchDurationMilliseconds = 700;
+    this.runner.matchCountdownMilliseconds = 0;
+    this.runner.startMatchCountdown();
+    setTimeout(() => {
```
Comment (shadedred) [inline tests/lib/services/gulag-runner.test.js:191]: https://github.com/destinygg/chat-bot/blob/master/tests/lib/services/punishment-cache.test.js


Don't use timeouts for tests, instead mock the clock using sinon, and control the advancement of it. The test I linked has an example of his this is done.

**Comment by shadedred** [comment]
AFter the timeouts are fixed, I'm good to merge this!

File: tests/lib/services/gulag-runner.test.js (hunk 1 above)
Comment (derDeidra) [reply tests/lib/services/gulag-runner.test.js:191]: Sure will do

**Comment by shadedred** [comment]
@ILiedAboutCake :shipit:

**Pull request merged**